File: ios/Classes/PlayerController.swift
import Foundation
import AVFoundation
import Flutter

class PlayerController: NSObject {
    var player: AVPlayer? { // Make player accessible but still internally managed
        return _player
    }
    private var _player: AVPlayer?
    private var timeObserverToken: Any?
    private let methodChannel: FlutterMethodChannel
    private let registrar: FlutterPluginRegistrar
    private var isPositionUpdaterRunning = false
    private let updateInterval: TimeInterval = 0.1 // 100ms same as Android
    private var currentVolume: Float = 0
    private var playbackSpeed: Float = 1.0
    
    init(methodChannel: FlutterMethodChannel, registrar: FlutterPluginRegistrar) {
        self.methodChannel = methodChannel
        self.registrar = registrar
        super.init()
        setupPlayer()
        populatePlayerState()
        setupMethodCallHandler()
    }
    
    
    private func setupPlayer() {
        // Initialize player if needed
        if _player == nil {
            _player = AVPlayer()
            _player?.actionAtItemEnd = .pause
            
            // Add rate observer to handle playback state changes
            _player?.addObserver(self, forKeyPath: "rate", options: [.new, .old], context: nil)
        }
        setupPlayerObservers()
    }
    
    private func populatePlayerState() {
        currentVolume = getPlayerVolume()
    }
    
    private func setupPlayerObservers() {
        // Remove any existing observers first
        if let timeObserverToken = timeObserverToken {
            _player?.removeTimeObserver(timeObserverToken)
            self.timeObserverToken = nil
        }
        
        // Time observer for position updates
        let interval = CMTime(seconds: updateInterval, preferredTimescale: CMTimeScale(NSEC_PER_SEC))
        timeObserverToken = _player?.addPeriodicTimeObserver(forInterval: interval, queue: .main) { [weak self] time in
            self?.sendCurrentPosition()
        }
        
        // Status observer for player
        _player?.addObserver(self, forKeyPath: "status", options: [.new, .old], context: nil)
        
        // Status observer for player item
        if let currentItem = _player?.currentItem {
            currentItem.addObserver(self, forKeyPath: "status", options: [.new, .old], context: nil)
        }
        
        // Rate observer for play/pause state
        _player?.addObserver(self, forKeyPath: "rate", options: [.new, .old], context: nil)
        
        // Add notification observer for when item finishes playing
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(playerItemDidReachEnd),
                                               name: .AVPlayerItemDidPlayToEndTime,
                                               object: nil)
    }
    
    private func handleAVError(_ error: NSError) -> [String: Any] {
        pauseVideo()
        let errorCode: Int
        
        // First check if it's a network error from URLError
        if error.domain == NSURLErrorDomain {
            print("Error Code: \(error.code)")
            switch error.code {
            case NSURLErrorNotConnectedToInternet, NSURLErrorNetworkConnectionLost, NSURLErrorTimedOut:
                errorCode = 1003 // networkError
            case NSURLErrorUnsupportedURL, NSURLErrorCannotFindHost, NSURLErrorFileDoesNotExist:
                errorCode = 1001 // sourceNotFound
            default:
                errorCode = 1000 // unknown
            }
        }
        // Then check AVFoundation errors
        else if error.domain == AVFoundationErrorDomain {
            if let avErrorCode = AVError.Code(rawValue: error.code) {
                switch avErrorCode {
                case .formatUnsupported, .decoderNotFound:
                    errorCode = 1008 // unsupportedFormat
                case .contentIsProtected:
                    errorCode = 1005 // drmError
                case .noLongerPlayable:
                    errorCode = 1006 // internalError
                case .mediaServicesWereReset:
                    errorCode = 1006 // internalError
                default:
                    errorCode = 1000
                }
            } else {
                errorCode = 1000 // unknown
            }
        } else {
            errorCode = 1000 // unknown
        }
        
        return [
            Constants.ERROR_CODE: errorCode,
            Constants.STACK_TRACE: "\(error.debugDescription)\nDomain: \(error.domain)\nCode: \(error.code)"
        ]
    }
    
    override func observeValue(forKeyPath keyPath: String?,
                               of object: Any?,
                               change: [NSKeyValueChangeKey : Any]?,
                               context: UnsafeMutableRawPointer?) {
        if keyPath == "status" {
            if object as? AVPlayer == _player {
                // Handle player status
                switch _player?.status {
                case .failed:
                    if let error = _player?.error as NSError? {
                        let errorDetails = handleAVError(error)
                        methodChannel.invokeMethod(Constants.ON_PLAYBACK_ERROR, arguments: errorDetails)
                    }
                case .unknown:
                    methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, arguments: PlaybackState.idle.rawValue)
                default:
                    break
                }
            } else if  object as? AVPlayerItem == _player?.currentItem {
                // Handle player item status
                if _player?.currentItem?.status == .readyToPlay {
                    methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, arguments: PlaybackState.ready.rawValue)
                    if let videoSize = getVideoSize(), videoSize.width > 0, videoSize.height > 0 {
                        let videoSizeData: [String: Any] = [
                            Constants.WIDTH: Int(videoSize.width),
                            Constants.HEIGHT: Int(videoSize.height)
                        ]
                        methodChannel.invokeMethod(Constants.ON_VIDEO_SIZE_CHANGED, arguments: videoSizeData)
                    }
                    sendDuration()
                } else if _player?.currentItem?.status == .failed {
                    if let error = _player?.currentItem?.error as NSError? {
                        let errorDetails = handleAVError(error)
                        methodChannel.invokeMethod(Constants.ON_PLAYBACK_ERROR, arguments: errorDetails)
                    }
                    methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, arguments: PlaybackState.idle.rawValue)
                }
            }
        } else if keyPath == "rate" {
            if let player = _player {
                let isPlaying = player.rate != 0
                methodChannel.invokeMethod(Constants.IS_PLAYING, arguments: isPlaying)
            }
        }
    }
    
    @objc func playerItemDidReachEnd(notification: NSNotification) {
        methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, arguments: PlaybackState.ended.rawValue)
    }
    
    private func setupMethodCallHandler() {
        methodChannel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }
            
            switch call.method {
            case Constants.INIT_PLAYER:
                if let args = call.arguments as? [String: Any],
                   let uri = args[Constants.URI] as? String,
                   let type = args[Constants.TYPE] as? String {
                    
                    let autoplay = args[Constants.AUTOPLAY] as? Bool ?? true
                    let enableMediaSession = args[Constants.ENABLE_MEDIA_SESSION] as? Bool ?? false
                    print("PLAYING_URI: \(uri)")
                    
                    self.initPlayer(uri: uri, type: type, autoplay: autoplay, enableMediaSession: enableMediaSession)
                    
                    result(nil)  // Send back response if needed
                } else {
                    result(FlutterError(code: Constants.INVALID_ARGUMENTS_CODE,
                                        message: Constants.MISSING_INVALID_ARGUMENTS_INIT_PLAYER,
                                        details: nil))
                }
            case Constants.SET_MEDIA_ITEM:
                if let args = call.arguments as? [String: Any],
                   let dataSource = args[Constants.DATA_SOURCE] as? [String: Any],
                   let type = dataSource[Constants.TYPE] as? String,
                   let uri = dataSource[Constants.URI] as? String {
                    
                    print("PLAYING_URI: \(uri)")
                    
                    let autoplay = args[Constants.AUTOPLAY] as? Bool ?? false
                    self.setMediaItemIfNeeded(uri: uri, type: type, autoplay: autoplay)
                    
                    result(nil) // Return success response
                } else {
                    result(FlutterError(code: Constants.INVALID_ARGUMENTS_CODE,
                                        message: Constants.MISSING_INVALID_ARGUMENTS_MEDIAITEM_ERROR,
                                        details: nil))
                }
            case Constants.PLAY_PAUSE:
                self.playPause()
                result(nil)
                
            case Constants.PLAY:
                self.playVideo()
                result(nil)
                
            case Constants.PAUSE:
                self.pauseVideo()
                result(nil)
                
            case Constants.GET_CURRENT_POSITION:
                result(player?.currentItem?.duration ?? 0)
                
            case Constants.STOP:
                self.stop()
                result(nil)
                
            case Constants.SEEK_TO:
                if let position = call.arguments as? Int {
                    self.seekTo(position: position)
                    result(nil)
                } else {
                    result(FlutterError(code: Constants.INVALID_ARGUMENTS_CODE,
                                        message: Constants.INVALID_POSITION_FOR_SEEK_TO,
                                        details: nil))
                }
            case Constants.MUTED:
                let muted = call.arguments as? Bool ?? false
                self.toggleMute(muted: muted)
                result(nil)
            case Constants.PLAYBACK_SPEED:
                let playbackSpeed = call.arguments as? Double ?? 1.0
                self.setPlaybackSpeed(playbackSpeed: playbackSpeed)
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    func initPlayer(uri: String?, type: String?, autoplay: Bool, enableMediaSession: Bool = false) {
        if (uri != nil && !uri!.isEmpty) {
            setMediaItemIfNeeded(uri: uri!, type: type, autoplay: autoplay)
        }
        
        if (player != nil) {
            methodChannel.invokeMethod(Constants.IS_INITIALIZED, arguments: true)
            if (enableMediaSession) {
                createMediaSession()
            }
        }
    }
    
    func createMediaSession() {
        let audioSession = AVAudioSession.sharedInstance()
        do {
            // Set the audio session category and mode.
            try audioSession.setCategory(.playback, mode: .moviePlayback)
            try audioSession.setActive(true)
        } catch {
            print(Constants.FAILED_TO_SET_AUDIO_SESSION_CONFIG)
        }
    }
    
    func autoplayMode(autoplay: Bool) {
        if autoplay {
            _player?.play()
            _player?.rate = self.playbackSpeed
        } else {
            if let isPlaying = _player?.timeControlStatus, isPlaying == .playing {
                pauseVideo()
            }
        }
    }
    
    func setMediaItemIfNeeded(uri: String, type: String?, autoplay: Bool = false) {
        let dataSource: DataSource
        
        switch type {
        case Constants.ASSET:
            do {
                dataSource = try DataSource.asset(context: registrar, assetPath: uri)
            } catch {
                print("Error loading asset: \(error.localizedDescription)")
                return
            }
        case Constants.NETWORK:
            guard let url = URL(string: uri), UIApplication.shared.canOpenURL(url) else {
                let errorDict = handleAVError(NSError(domain: NSURLErrorDomain, code: NSURLErrorUnsupportedURL, userInfo: nil))
                methodChannel.invokeMethod(Constants.ON_PLAYBACK_ERROR, arguments: errorDict)
                return
            }
            dataSource = DataSource.network(url: uri)
        case Constants.FILE:
            dataSource = DataSource.file(filePath: uri)
        default:
            print(Constants.UNSUPPORTED_URI_TYPE)
            return
        }
        
        let newMediaAsset = dataSource.getMediaAsset()
        
//        // Remove observer from old item if exists
//        if let currentItem = _player?.currentItem {
//            currentItem.removeObserver(self, forKeyPath: "status")
//        }
//        
//        // Check if the player is already playing the same media item
//        if let currentItem = _player?.currentItem,
//           currentItem.asset.isEqual(newMediaAsset) {
//            currentItem.addObserver(self, forKeyPath: "status", options: [.new, .old], context: nil)
//            return // Already playing the same media
//        }
        
        let playerItem = AVPlayerItem(asset: newMediaAsset)
        playerItem.addObserver(self, forKeyPath: "status", options: [.new, .old], context: nil)
        _player?.replaceCurrentItem(with: playerItem)
        
        methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, arguments: PlaybackState.loading.rawValue)
        
        autoplayMode(autoplay: autoplay)
    }
    
    private func playPause() {
        if let player = _player {
            if player.rate != 0 {
                pauseVideo()
            } else {
                playVideo()
            }
        }
    }
    
    private func playVideo() {
        if let player = _player, player.currentItem?.status == .readyToPlay {
            if player.currentTime() >= player.currentItem?.duration ?? .zero {
                player.seek(to: .zero)
            }
            player.play()
            player.rate = self.playbackSpeed
        }
    }
    
    private func pauseVideo() {
        _player?.pause()
    }
    
    private func stop() {
        pauseVideo()
        _player?.seek(to: .zero)
    }
    
    private func seekTo(position: Int) {
        let time = CMTime(seconds: Double(position) / 1000.0, preferredTimescale: CMTimeScale(NSEC_PER_SEC))
        _player?.seek(to: time)
    }
    
    private func getCurrentPosition() -> Float64 {
        let currentPosition = CMTimeGetSeconds(player?.currentTime() ?? CMTime.zero)
        return currentPosition
    }
    
    private func getDuration() -> Float64 {
        guard let item = _player?.currentItem,
              item.status == .readyToPlay,
              CMTIME_IS_NUMERIC(item.duration) else { return 0 }
        
        let duration = CMTimeGetSeconds(item.duration)
        return duration.isFinite ? duration : 0
    }

    
    private func getBufferedPosition() -> Float64 {
        guard let timeRange = player?.currentItem?.loadedTimeRanges.first?.timeRangeValue else { return 0 }
        let bufferedPosition = CMTimeGetSeconds(timeRange.start) + CMTimeGetSeconds(timeRange.duration)
        return bufferedPosition
    }
    
    private func sendCurrentPosition() {
        guard let player = _player, let currentItem = player.currentItem else { return }
        
        let currentTime = getCurrentPosition()
        let bufferedTime = getBufferedPosition()
        
        // Check for valid values before converting to Int
        if currentTime.isFinite {
            let position = Int(currentTime * 1000)
            methodChannel.invokeMethod(Constants.CURRENT_POSITION, arguments: position)
        }
        
        if bufferedTime.isFinite {
            let buffered = Int(bufferedTime * 1000)
            methodChannel.invokeMethod(Constants.BUFFERED_POSITION, arguments: buffered)
        }
    }
    
    private func sendDuration() {
        let durationSeconds = getDuration()
        if durationSeconds.isFinite {
            methodChannel.invokeMethod(Constants.DURATION, arguments: Int(durationSeconds * 1000))
        }
    }
    
    func releasePlayer() {
        if let timeObserverToken = timeObserverToken {
            _player?.removeTimeObserver(timeObserverToken)
        }
        if let currentItem = _player?.currentItem {
            currentItem.removeObserver(self, forKeyPath: "status")
        }
        _player?.removeObserver(self, forKeyPath: "status")
        _player?.removeObserver(self, forKeyPath: "rate")
        NotificationCenter.default.removeObserver(self)
        _player = nil
    }
    
    func toggleMute(muted: Bool) {
        if muted {
            currentVolume = getPlayerVolume()
            setPlayerVolume(0)
        } else {
            setPlayerVolume(currentVolume)
        }
    }

    func setPlaybackSpeed(playbackSpeed: Double) {
        if playbackSpeed > 0 {
            self.playbackSpeed = Float(playbackSpeed)
            player?.rate = self.playbackSpeed
        }
    }
    
    func getPlayerVolume() -> Float {
        return player?.volume ?? 0
    }
    
    func setPlayerVolume(_ volume: Float) {
        player?.isMuted = volume == 0
        player?.volume = volume
    }
    
    func getVideoSize() -> CGSize? {
        guard let track = player?.currentItem?.asset.tracks(withMediaType: .video).first else {
            return nil
        }
        let size = track.naturalSize.applying(track.preferredTransform)
        return CGSize(width: abs(size.width), height: abs(size.height))
    }
}

File: android/src/main/kotlin/com/gvuitech/mediax/PlayerController.kt
package com.gvuitech.mediax

import android.annotation.SuppressLint
import android.app.PictureInPictureParams
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore.Video
import android.util.Log
import android.util.Rational
import androidx.media3.common.ForwardingPlayer
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.VideoSize
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.session.MediaSession
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

class PlayerController(
    private val activity: Context,
    flutterEngine: FlutterEngine,
    private val controllerId: String,
    private val methodChannel: MethodChannel
) {

    private var player: Player? = null
    private var mediaSession: MediaSession? = null

    private val updateInterval = 100L
    private val handler = Handler(Looper.getMainLooper())
    private var isPositionUpdaterRunning = false

    private var currentVolume = 0f

    init {
        getPlayer(activity)
        populatePlayerState()
        setupPlayerListeners()
        setupNativeCalls()
    }

    fun getPlayer(context: Context): Player {
        if (player == null) {
            player = ExoPlayer.Builder(context).build()
        }
        return player!!
    }

    private fun populatePlayerState() {
        this.currentVolume = getPlayerVolume()
    }

    private fun setMediaItemIfNeeded(uri: String, type: String?, autoplay: Boolean = false) {
        val dataSource = when (type) {
            Constants.ASSET -> DataSource.asset(activity, uri)
            Constants.NETWORK -> DataSource.network(uri)
            Constants.FILE -> DataSource.file(uri)
            else -> throw IllegalArgumentException(Constants.UNSUPPORTED_URI_TYPE)
        }

        player?.setMediaItem(dataSource.getMediaItem())

        player?.prepare()

        autoplayMode(autoplay)
    }

    fun initPlayer(uri: String?, type: String?, autoplay: Boolean = true, enableMediaSession: Boolean = false) {
        if (!uri.isNullOrEmpty()) {
            setMediaItemIfNeeded(uri, type, autoplay)
        }

        if (player != null) {
            methodChannel.invokeMethod(Constants.IS_INITIALIZED, true)
            if (enableMediaSession) {
                createMediaSession()
            }
        }
    }

    private fun createMediaSession() {
        val forwardingPlayer = @UnstableApi object : ForwardingPlayer(player!!) {
            override fun getAvailableCommands(): Player.Commands {
                return super.getAvailableCommands()
                    .buildUpon()
                    .remove(COMMAND_SEEK_TO_PREVIOUS)
                    .remove(COMMAND_SEEK_TO_NEXT)
                    .build()
            }
        }
        mediaSession = MediaSession.Builder(activity.applicationContext, forwardingPlayer).build()
    }

    private fun autoplayMode(autoplay: Boolean) {
        if (autoplay) {
            player?.play()
        } else {
            if (player?.isPlaying == true) pauseVideo()
        }
    }

    private fun playPause() {
        if (player?.isPlaying == true) {
            pauseVideo()
        } else {
            playVideo()
        }
    }

    private fun playVideo() {
        if (player?.playbackState == Player.STATE_ENDED) {
            player!!.seekTo(0)
        }
        player?.play()
    }

    private fun pauseVideo() {
        player?.pause()
    }

    private fun stop() {
        player?.stop()
    }

    fun releasePlayer() {
        handler.removeCallbacksAndMessages(null)
        methodChannel.setMethodCallHandler(null)
        player?.release()
        mediaSession?.release()
        mediaSession = null
        player = null
    }

    private fun seekTo(arguments: Any?) {
        val position = arguments ?: 0L
        player?.seekTo(position.toString().toLong())
        sendCurrentPosition()
    }

    private fun getCurrentPosition(): Long {
        return player?.currentPosition ?: 0L
    }

    private fun getBufferedPosition(): Long {
        return player?.bufferedPosition ?: 0L
    }

    private fun getDuration(): Long {
        return player?.duration ?: 0L
    }

    fun startPositionUpdates() {
        if (isPositionUpdaterRunning) return

        isPositionUpdaterRunning = true
        handler.postDelayed(object : Runnable {
            override fun run() {
                sendCurrentPosition()
                handler.postDelayed(this, updateInterval)
            }
        }, updateInterval)
    }

    fun stopPositionUpdates() {
        handler.removeCallbacksAndMessages(null)
        isPositionUpdaterRunning = false
    }

    @SuppressLint("UnsafeOptInUsageError")
    fun enterPipMode() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val params = PictureInPictureParams.Builder()
                .setAspectRatio(
                    Rational(
                        player?.videoSize?.width.takeIf { (it ?: 0) > 0 } ?: 1280,
                        player?.videoSize?.height.takeIf { (it ?: 0) > 0 } ?: 720
                    )
                ).build()
//            activity.enterPictureInPictureMode(params)
        }
    }

    private fun exitPipMode() {

    }

    private fun setupPlayerListeners() {
        player?.addListener(object : Player.Listener {
            override fun onVideoSizeChanged(videoSize: VideoSize) {
                super.onVideoSizeChanged(videoSize)
                if (videoSize.width == 0 || videoSize.height == 0) return
                methodChannel.invokeMethod(Constants.ON_VIDEO_SIZE_CHANGED, mapOf(
                    Constants.WIDTH to (videoSize.width),
                    Constants.HEIGHT to (videoSize.height)
                ))
            }

            override fun onIsPlayingChanged(isPlaying: Boolean) {
                methodChannel.invokeMethod(Constants.IS_PLAYING, isPlaying)
                if (isPlaying) {
                    startPositionUpdates()
                } else {
                    stopPositionUpdates()
                }
            }

            override fun onPlayerError(error: PlaybackException) {
                super.onPlayerError(error)

                val errorCode = when (error.errorCode) {
                    PlaybackException.ERROR_CODE_IO_NETWORK_CONNECTION_FAILED,
                    PlaybackException.ERROR_CODE_IO_BAD_HTTP_STATUS,
                    PlaybackException.ERROR_CODE_IO_NETWORK_CONNECTION_TIMEOUT -> 1003 // networkError
                    PlaybackException.ERROR_CODE_IO_FILE_NOT_FOUND -> 1001 // sourceNotFound
                    PlaybackException.ERROR_CODE_DECODER_INIT_FAILED,
                    PlaybackException.ERROR_CODE_AUDIO_TRACK_INIT_FAILED -> 1002 // codecError
                    PlaybackException.ERROR_CODE_TIMEOUT -> 1004 // timeout
                    PlaybackException.ERROR_CODE_DRM_SYSTEM_ERROR -> 1005 // drmError
                    PlaybackException.ERROR_CODE_UNSPECIFIED -> 1000 // unknown
                    else -> 1006 // internalError
                }

                val errorDetails = mapOf(
                    Constants.ERROR_CODE to errorCode,
                    Constants.STACK_TRACE to error.stackTraceToString()
                )
                methodChannel.invokeMethod(Constants.ON_PLAYBACK_ERROR, errorDetails)
            }

            override fun onIsLoadingChanged(isLoading: Boolean) {
                super.onIsLoadingChanged(isLoading)
                if (player != null) {
                    if (isLoading) {
                        sendPlaybackState(PlaybackState.loading)
                    } else {
                        parsePlaybackState(player!!.playbackState)
                    }
                }
            }

            override fun onPlaybackStateChanged(playbackState: Int) {
                super.onPlaybackStateChanged(playbackState)
                parsePlaybackState(playbackState)
                when (playbackState) {
                    Player.STATE_IDLE -> {
                        seekTo(0)
                        stopPositionUpdates()
                    }
                    Player.STATE_READY -> {
                        sendDuration()
                        if (player?.isPlaying == true) {
                            startPositionUpdates()
                        }
                    }
                    Player.STATE_ENDED -> {
                        stopPositionUpdates()
                    }

                    Player.STATE_BUFFERING -> {
                    }
                }
            }
        })
    }

    private fun setupNativeCalls() {
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                Constants.INIT_PLAYER -> {
                    val args = call.arguments as Map<*, *>
                    val uri = args[Constants.URI] as String
                    val type = args[Constants.TYPE] as String
                    val autoplay = args[Constants.AUTOPLAY] as Boolean? ?: true
                    val enableMediaSession = args[Constants.ENABLE_MEDIA_SESSION] as Boolean? ?: false
                    initPlayer(uri, type, autoplay, enableMediaSession)
                }
                Constants.SET_MEDIA_ITEM -> {
                    val args = call.arguments as Map<*, *>

                    val dataSource = args[Constants.DATA_SOURCE] as? Map<*, *>
                        ?: throw IllegalArgumentException(Constants.DATASOURCE_NULL_ERROR)

                    val type = dataSource[Constants.TYPE] as? String ?: throw IllegalArgumentException(Constants.DATASOURCE_TYPE_NULL_ERROR)

                    val uri = dataSource[Constants.URI] as? String ?: throw IllegalArgumentException(Constants.URI_NULL_ERROR)

                    val autoplay = args[Constants.AUTOPLAY] as? Boolean ?: false

                    setMediaItemIfNeeded(uri, type, autoplay)
                }
                Constants.PLAY_PAUSE -> playPause()
                Constants.PLAY -> playVideo()
                Constants.PAUSE -> pauseVideo()
                Constants.GET_CURRENT_POSITION -> result.success(getCurrentPosition())
                Constants.GET_DURATION -> result.success(getDuration())
                Constants.SEEK_TO -> seekTo(call.arguments)
                Constants.ENTER_PIP -> enterPipMode()
                Constants.EXIT_PIP -> exitPipMode()
                Constants.STOP -> stop()
                Constants.RELEASE_PLAYER -> releasePlayer()
                Constants.MUTED -> {
                    val muted = call.arguments as? Boolean ?: false
                    toggleMute(muted)
                }
                Constants.PLAYBACK_SPEED -> {
                    val playbackSpeed = call.arguments as? Double ?: 1.0
                    setPlaybackSpeed(playbackSpeed)
                }
                else -> {

                }
            }
        }
    }

    private fun setPlaybackSpeed(playbackSpeed: Double) {
        if (playbackSpeed > 0) {
            player?.setPlaybackSpeed(playbackSpeed.toFloat())
        }
    }

    private fun toggleMute(muted: Boolean) {
        if (muted) {
            this.currentVolume = getPlayerVolume()
            setPlayerVolume(0f)
        } else {
            setPlayerVolume(1f)
        }
    }

    private fun getPlayerVolume(): Float {
        return player?.volume ?: 0f
    }

    private fun setPlayerVolume(volume: Float) {
        player?.volume = volume
    }

    private fun parsePlaybackState(state: Int) {
        var currentState: PlaybackState? = null
        when (state) {
            Player.STATE_IDLE -> currentState = PlaybackState.idle
            Player.STATE_READY -> currentState = PlaybackState.ready
            Player.STATE_BUFFERING -> currentState = PlaybackState.buffering
            Player.STATE_ENDED -> currentState = PlaybackState.ended
        }

        if (currentState != null) {
            sendPlaybackState(currentState)
        }
    }

    private fun sendPlaybackState(playbackState: PlaybackState) {
        methodChannel.invokeMethod(Constants.ON_PLAYBACK_STATE_CHANGED, playbackState.ordinal)
    }

    private fun sendDuration() {
        methodChannel.invokeMethod(Constants.DURATION, getDuration())
    }

    private fun sendCurrentPosition() {
        val currentPosition = getCurrentPosition()
        val bufferedPosition = getBufferedPosition()

        // Send current position along with playback speed to Flutter
        methodChannel.invokeMethod(Constants.CURRENT_POSITION, currentPosition)
        methodChannel.invokeMethod(Constants.BUFFERED_POSITION, bufferedPosition)
    }



}